File: base/004/solver.cpp
#include <iostream>
#include <cstdio>
#include <clocale>
#include <sstream>
#include <iomanip> //setprecision

using namespace std;

struct Grafite{
    float calibre;
    string dureza;
    int tamanho;

    Grafite(float calibre, string dureza, int tamanho){
        this->calibre = calibre;
        this->dureza = dureza;
        this->tamanho = tamanho;
    }
    friend ostream& operator<<(ostream& os, Grafite g){
        os << std::fixed << setprecision(1) << g.calibre << ":" << g.dureza << ":" << g.tamanho;
        return os;
    }
    int desgastePorFolha(){
        if(dureza == "HB")
            return 1;
        else if(dureza == "2B")
            return 2;
        else if(dureza == "4B")
            return 4;
        else
            return 6;
    }
};

struct Lapiseira{
    float calibre;
    Grafite * grafite;

    Lapiseira(float calibre = 0.0){
        this->calibre = calibre;
        this->grafite = nullptr;
    }
    ~Lapiseira(){
        if(this->grafite != nullptr)
            delete this->grafite;
    }

    friend ostream& operator<<(ostream& os, const Lapiseira& l){
        os << "calibre: " << l.calibre << ", grafite: ";
        if (l.grafite != nullptr)
            os << "[" << *l.grafite << "]";
        else
            os << "null";
        return os;
    }

    bool inserir(Grafite * grafite){
        if(this->grafite == nullptr){
            if(this->calibre != grafite->calibre){
                cout << "fail: calibre incompatível\n";
                return false;
            }else{
                this->grafite = grafite;
                return true;
            }
        }
        cout << "fail: ja existe grafite\n";
        return false;
    }

    Grafite * remover(){
        if(this->grafite == nullptr){
            cout << "fail: nao existe grafite\n";
            return nullptr;
        }
        auto backup = this->grafite;
        this->grafite = nullptr;
        return backup;
    }

    void write(int folhas){
        if(this->grafite == nullptr){
            cout << "fail: nao existe grafite\n";
            return;
        }
        int& tamanho = this->grafite->tamanho;
        int total = this->grafite->desgastePorFolha() * folhas;
        if(tamanho > total){
            tamanho -= total;
            return;
        }
        if(tamanho < total){
            int qtd = tamanho / this->grafite->desgastePorFolha();
            cout << "fail: folhas escritas completas: " << qtd << "\n";
        }
        cout << "warning: grafite acabou\n";
        delete this->grafite;
        this->grafite = nullptr;
    }
};

int main(){
    Lapiseira lapiseira;
    string line;
    while(true){
        getline(cin, line);
        stringstream ss(line);
        string cmd;
        cout << "$" << line << endl;
        ss >> cmd;
        if(line == "end")
            break;
        else if(cmd == "help"){
            cout << "init _calibre; inserir _calibre _dureza _tamanho; remover; write _folhas" << endl;
        }else if(cmd == "init"){
            float calibre;
            ss >> calibre;
            lapiseira = Lapiseira(calibre);
        }else if(cmd == "inserir"){
            float calibre;
            string dureza;
            int tamanho;
            ss >> calibre >> dureza >> tamanho;
            auto grafite = new Grafite(calibre, dureza, tamanho);
            if(!lapiseira.inserir(grafite))
                delete grafite;
        }else if(cmd == "remover"){
            auto gr = lapiseira.remover();
            if(gr != nullptr)
                delete gr;
        }else if(cmd == "show"){
            cout << lapiseira << endl;
        }else if (cmd == "write"){
            int pressao;
            ss >> pressao;
            lapiseira.write(pressao);
        }else{
            cout << "fail: comando invalido" << endl;
        }
    }
    return 0;
}

File: base/004/Solver.java
import java.text.DecimalFormat;
import java.util.Scanner;

class Grafite {
    public float calibre;
    public String dureza;
    public int tamanho;

    public Grafite(float calibre, String dureza, int tamanho) {
        this.calibre = calibre;
        this.dureza = dureza;
        this.tamanho = tamanho;
    }
    public String toString() {
        DecimalFormat form = new DecimalFormat("0.0");
        return form.format(calibre) + ":" + dureza + ":" + tamanho;
    }
    public int desgastePorFolha() {
        if(dureza.equals("HB"))
            return 1;
        else if(dureza.equals("2B"))
            return 2;
        else if(dureza.equals("4B"))
            return 4;
        else
            return 6;
    }
}

class Lapiseira {
    public float calibre;
    public Grafite grafite;

    public Lapiseira(float calibre) {
        this.calibre = calibre;
        this.grafite = null;
    }

    public String toString() {
        String saida = "calibre: " + calibre + ", grafite: ";
        if (grafite != null)
            saida += "[" + grafite + "]";
        else
            saida += "null";
        return saida;
    }

    public boolean inserir(Grafite grafite) {
        if(this.grafite == null) {
            if(this.calibre != grafite.calibre) {
                System.out.println("fail: calibre incompatível");
                return false;
            } else {
                this.grafite = grafite;
                return true;
            }
        }
        System.out.println("fail: ja existe grafite");
        return false;
    }

    public Grafite remover() {
        if(this.grafite == null) {
            System.out.println("fail: nao existe grafite");
            return null;
        }
        Grafite backup = this.grafite;
        this.grafite = null;
        return backup;
    }

    public void write(int folhas) {
        if(this.grafite == null) {
            System.out.println("fail: nao existe grafite");
            return;
        }
        this.grafite.tamanho -= this.grafite.desgastePorFolha() * folhas;
        if(this.grafite.tamanho < 0) {
            System.out.println("fail: folhas escritas completas: " + (folhas + (int)(this.grafite.tamanho/this.grafite.desgastePorFolha()) - 1));
        }
        if(this.grafite.tamanho <= 0) {
            System.out.println("warning: grafite acabou");
            this.grafite = null;
        }
    }
}
//!KEEP
class Solver{
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Lapiseira lapiseira = new Lapiseira(0.5f);
        while(true) {
            String line = scanner.nextLine();
            System.out.println("$" + line);
            String ui[] = line.split(" ");
            if(ui[0].equals("end")) {
                break;
            } else if(ui[0].equals("help")) {
                System.out.println("init _calibre; inserir _calibre _dureza _tamanho; remover; write _folhas");
            } else if(ui[0].equals("init")) { //calibre
                lapiseira = new Lapiseira(Float.parseFloat(ui[1]));
            } else if(ui[0].equals("inserir")) {//calibre dureza tamanho
                float calibre = Float.parseFloat(ui[1]);
                String dureza  = ui[2];
                int tamanho = Integer.parseInt(ui[3]);
                lapiseira.inserir(new Grafite(calibre, dureza, tamanho));
            } else if(ui[0].equals("remover")) {
                lapiseira.remover();
            } else if(ui[0].equals("show")) {
                System.out.println(lapiseira);
            } else if (ui[0].equals("write")) {
                lapiseira.write(Integer.parseInt(ui[1]));
            } else {
                System.out.println("fail: comando invalido");
            }
        }
    }
}

class Manual {
    public static void main(String[] args) {
        //case inserindo grafites
        Lapiseira lapiseira = new Lapiseira(0.5f);
        System.out.println(lapiseira);
        //calibre: 0.5, grafite: null
        lapiseira.inserir(new Grafite(0.7f, "2B", 50));
        //fail: calibre incompatível
        lapiseira.inserir(new Grafite(0.5f, "2B", 50));
        System.out.println(lapiseira);
        //calibre: 0.5, grafite: [0.5:2B:50]

        //case inserindo e removendo
        lapiseira = new Lapiseira(0.3f);
        lapiseira.inserir(new Grafite(0.3f, "2B", 50));
        System.out.println(lapiseira);
        //calibre: 0.3, grafite: [0.3:2B:50]
        lapiseira.inserir(new Grafite(0.3f, "4B", 70));
        //fail: ja existe grafite
        System.out.println(lapiseira);
        //calibre: 0.3, grafite: [0.3:2B:50]
        lapiseira.remover();
        lapiseira.inserir(new Grafite(0.3f, "4B", 70));
        System.out.println(lapiseira);
        //calibre: 0.3, grafite: [0.3:4B:70]

        //case escrevendo 1
        lapiseira = new Lapiseira(0.9f);
        lapiseira.inserir(new Grafite(0.9f, "4B", 4));
        lapiseira.write(1);
        //warning: grafite acabou
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: null
        lapiseira.inserir(new Grafite(0.9f, "4B", 30));
        lapiseira.write(6);
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: [0.9:4B:6]
        lapiseira.write(3);
        //fail: folhas escritas completas: 1
        //warning: grafite acabou
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: null

        //case escrevendo 2
        lapiseira = new Lapiseira(0.9f);
        lapiseira.inserir(new Grafite(0.9f, "2B", 15));
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: [0.9:2B:15]
        lapiseira.write(4);
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: [0.9:2B:7]
        lapiseira.write(4);
        //fail: folhas escritas completas: 3
        //warning: grafite acabou
        System.out.println(lapiseira);
        //calibre: 0.9, grafite: null
    }
}
//!OFF